### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SCREEN READER, COGNITION]: Error messages MUST be associated with that input

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#16222]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:** Jennifer

## User Story or Problem Statement

As a screen reader user, I expect to see error messages paired with the input that has the error in a consistent treatment throughout the application or website.

## Details

On Step 2 of 4: Contested issues and "Select the issue you would like to have reviewed (*Required)", if a user does not select a disability, the error message is presented in an alertbox beneath the checkbox options. The error here is on the label, "Select the issue you would like to have reviewed (*Required)" rather than on the checkboxes themselves, so the error message **MUST** appear beneath the label, to be consistent with error handling across VA.gov. Please see screenshots.

Additionally, **consider** adding an error icon preceding error messages. This would be a design system proposal for a change to existing error message styling. For VSA to propose an update, we must:

1. Discuss the problem found amongst VSA teams
2. Explore and document actions amongst VSA teams
3. Document a solution that works for VSA teams
4. Propose to VSP design system council for review

So, if this suggestion of adding the icon sounds like something your team wants to explore, please work with your accessibility specialists and design lead to follow the above steps. 

## Acceptance Criteria

- [ ] Error message appears beneath the `<legend>Select the issue you would like to have reviewed(*Required)</legend>
- [ ] Error message is announced when the user attempts to continue without making a selection 

## Environment

* Operating System: all
* Browser: all
* Screenreading device: all
* Server destination: staging

## Steps to Recreate

1. Enter `https://staging.va.gov/decision-reviews/higher-level-review/request-higher-level-review-form-20-0996/contested-issues` in browser
2. Navigate to Step 2 of 4: Contested issues
3. Submit the screen without selecting an option
4. Verify that the error message is presented in an alertbox beneath the checkbox card

### Current

```diff
<form class="rjsf" novalidate="">
  <fieldset>
    <div class="input-section schemaform-field-container">
      <div class="schemaform-block-header">
!        <strong>Select the issue you would like to have reviewed</strong><span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</span>
      </div>
      <div>
        <div class="schemaform-field-template usa-input-error schemaform-first-field" tabindex="-1">
!          <label class="usa-input-error-label schemaform-label" for="root_contestedIssues" id="root_contestedIssues-label"><span class="schemaform-required-span">(*Required)</span></label><span class="usa-input-error-message" id="root_contestedIssues-error-message" role="alert"><span class="sr-only">Error</span> Please select a contested issue</span>
          <div class="schemaform-widget-wrapper">
            <dl class="review">
              <dt class="form-checkbox widget-outline">
                <input id="root_contestedIssues_0" name="root_contestedIssues_0" required="" type="checkbox">
!               <label class="schemaform-label" for="root_contestedIssues_0"></label>
              </dt>
              <dd>
                <div class="widget-content">
!                  <label class="schemaform-label" for="root_contestedIssues_0"></label>
!                  <h3 class="vads-u-margin-y--0 vads-u-font-size--h4">
!                    <label class="schemaform-label" for="root_contestedIssues_0">Hearing loss</label>
!                  </h3>
!                  <p class="vads-u-margin-bottom--0"><label class="schemaform-label" for="root_contestedIssues_0">Service connection for Hearing loss is granted with an evaluation of 10 percent effective July 15, 2018.</label></p>
!                  <p><label class="schemaform-label" for="root_contestedIssues_0">Decision date: <strong>June 15, 2020</strong></label></p>
                </div>
              </dd>
!             <dd></dd>
            </dl>
          </div>
          <div></div>
        </div>
      </div>
      <div>
        <div>
          <div class="schemaform-field-container">
            <div class="schemaform-block-header">
!             <div class="usa-alert usa-alert-error contested-issues-error vads-u-margin-top--1">
!               <div class="usa-alert-body">
!                  <h3 class="usa-alert-heading">Please choose a disability</h3>
!                  <div class="usa-alert-text">
!                   You need to choose the rated disability you’re requesting for a Higher-Level Review. We can’t process your request without a selected disability.
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
```

### Recommendation
```diff
<form class="rjsf" novalidate="">
!  <fieldset>
    <div class="input-section schemaform-field-container">
      <div class="schemaform-block-header">
<!-- change the <strong> label to a <legend> in another issue ticket https://github.com/department-of-veterans-affairs/va.gov-team/issues/16224 -->
+        <legend>Select the issue you would like to have reviewed<span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</span></legend>
      </div>
      <div>
        <div class="schemaform-field-template usa-input-error schemaform-first-field" tabindex="-1">
<!-- I'm not sure if this label is necessary, maybe it is for the VAFS -->
!         <label class="usa-input-error-label schemaform-label" for="root_contestedIssues" id="root_contestedIssues-label"><span class="schemaform-required-span">(*Required)</span></label>
<!-- error message from the developer design system documentation https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/visual-design/components/errorablecheckboxgroup/ -->
+        <span class="usa-input-error-message" role="alert" id="default-error-message">
+          <span class="sr-only">Error:</span> 
+          Please choose a disability. You need to choose the rated disability you’re requesting for a Higher-Level Review. We can’t process your request without a selected disability.</span>
<!-- checkbox card styling and behavior should be attached with the div.schemaform-widget-wrapper or the dl and the children within -->
          <div class="schemaform-widget-wrapper">
            <dl class="review">
<!-- structured Hearing loss checkbox and label in the dt -->
+              <dt class="form-checkbox widget-outline">
                  <input id="root_contestedIssues_0" name="root_contestedIssues_0" required="" type="checkbox">
                  <label class="schemaform-label" for="root_contestedIssues_0">
                    <span class="vads-u-margin-y--0 vads-u-font-size--h4">Hearing loss</span>
                  </label>
              </dt>
<!-- set the description of Hearing loss in a dd -->
+              <dd>
                <div class="widget-content">
                    <p class="vads-u-margin-bottom--0">Service connection for Hearing loss is granted with an evaluation of 10 percent effective July 15, 2018.</p>
                    <p class="schemaform-label"><dfn>Decision date:</dfn> <strong>June 15, 2020</strong></p>
                </div>
              </dd>
+ <!-- removed empty <dd> -->
            </dl>
          </div>
          <div></div>
        </div>
      </div>

```

## WCAG or Vendor Guidance (optional)

* [WCAG 2.0 Level AA - 3.3.3](https://www.wuhcag.com/error-suggestion/)

* [Adrian Roselli - Avoid Messages Under Fields](https://adrianroselli.com/2017/01/avoid-messages-under-fields.html)

* [error message from the developer design system documentation](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/visual-design/components/errorablecheckboxgroup/)

### Current 

![Screen Shot 2020-11-16 at 12 08 36 PM](https://user-images.githubusercontent.com/57469/99288087-16f67780-2809-11eb-8589-1a13bd53029b.png)

### Recommended

![image (14)](https://user-images.githubusercontent.com/57469/99288889-3e9a0f80-280a-11eb-9359-941ea3b221e2.png)

**Comment by Mottie** [comment]
Fixed in https://github.com/department-of-veterans-affairs/vets-website/pull/15349

**Comment by joshkimux** [comment]
@Mottie added a new acceptance criteria for the error to be read out upon attempting to continue. Also added `aria-describedby` to the fieldset to associate the error of not making a selection to the fieldset as a whole.

**Comment by Mottie** [comment]
Actually adding the `aria-describedby` attribute to the `fieldset` would require platform code changes 😿

**Comment by joshkimux** [comment]
@Mottie hmm, would it be possible to insert the error message within the legend itself?

**Comment by Mottie** [comment]
Yes

**Comment by joshkimux** [comment]
@Mottie Ignore my previous comment on inserting the error message within the legend. 

I've tested this again on NVDA/Firefox and VO/Safari and the error messages announces once the user attempts to continue without making a selection. This should be good to go, but would probably need a once over on JAWs as well just in case @1Copenut .
